### PR TITLE
[Reviewer: Alex] Allow construction of a localhost-only memcached

### PR DESCRIPTION
**Pull request by rkd-msw**
This adds a new MemcachedStore constructor, which just directs all traffic to 127.0.0.1:11211.

I've left all the rest of the code in although Sprout/Ralf/etc. no longer need much of it - this is because:

* you're going to need it for Astaire
* the more of this we rewrite, the more painful merges will be
* some of it is useful - e.g. trying twice to workaround a memcached connection reuse bug
* my use-case is genuinely a special case of the general use-case - I need to connect to one server with one vbucket and one replica

### CODE REVIEW

**Comment by rkd-msw** [comment]
By the way, I think I can merge this to master rather than keeping a long-running branch - shout if you disagree.

File: src/memcachedstore.cpp [hunk 1]
```diff
@@ -107,6 +107,44 @@ BaseMemcachedStore::BaseMemcachedStore(bool binary,
...
+
+  // Create the mutex for protecting vbucket comm state.
+  pthread_mutex_init(&_vbucket_comm_lock, NULL);
+
+  // Set up the fixed options for memcached.  We use a very short connect
+  // timeout because libmemcached tries to connect to all servers sequentially
+  // during start-up, and if any are not up we don't want to wait for any
+  // significant length of time.
+  _options = "--CONNECT-TIMEOUT=10 --SUPPORT-CAS --POLL-TIMEOUT=250";
+  _options += (_binary) ? " --BINARY-PROTOCOL" : "";
```
Comment (AlexHockey) [inline src/memcachedstore.cpp]: You initialize the `_vbucket_comm_lock` but not `_vbucket_comm_state`. Is this right?

Comment (rkd-msw) [inline src/memcachedstore.cpp]: Yep - we'll never touch `_vbucket_comm_state` if there's no vbucket alarm (and we don't want one here), but we'll destroy the lock in the destructor. I could make that conditional on the alarm existing, but it doesn't feel very valuable.

Comment (AlexHockey) [inline src/memcachedstore.cpp]: Quite a lot of code duplication here. Subfunction?

**Comment by AlexHockey** [comment]
Hmm.... I'm not a big fan of this change (sorry!). Some of the members are only initialized if the constructor was called in a certain way are are irrelevant otherwise. I think a better approach would be to define two subclasses of `BaseMemcachedStore` and move the functionality into whichever class actually needs it.

File: src/memcachedstore.cpp [hunk 2]
```diff
@@ -107,6 +107,44 @@ BaseMemcachedStore::BaseMemcachedStore(bool binary,
...
+}
+
```
Comment (AlexHockey) [inline src/memcachedstore.cpp]: Quite a lot of code duplication here. Subfunction?

**Comment by rkd-msw** [comment]
We agreed (after toying briefly with OmniscientMemcachedStore and MonoscientMemcachedStore) on Topology{Aware,Neutral}MemcachedStore.

File: src/memcachedstore.cpp (hunk 1 above)
Comment (AlexHockey) [inline src/memcachedstore.cpp]: If we're initializing things we never touch lets initialize everything. There is less chance of catching someone out in the future.

**Comment by rkd-msw** [comment]
Better?

File: src/memcachedstore.cpp [hunk 3]
```diff
@@ -1021,24 +1025,48 @@ void BaseMemcachedStore::log_delete_failure(const std::string& fqkey,
...
+{
+  set_fixed_server("127.0.0.1:11211");
+}
+
+void TopologyNeutralMemcachedStore::set_fixed_server(std::string server)
+{
+  _view_number++;
+  _servers = {server};
+  _read_replicas = {{server}};
+  _write_replicas = {{server}};
```
Comment (AlexHockey) [inline src/memcachedstore.cpp]: Wow, I had no idea you could do that!

File: src/memcachedstore.cpp [hunk 4]
```diff
@@ -97,8 +99,11 @@ BaseMemcachedStore::BaseMemcachedStore(bool binary,
   _options = "--CONNECT-TIMEOUT=10 --SUPPORT-CAS --POLL-TIMEOUT=250";
   _options += (_binary) ? " --BINARY-PROTOCOL" : "";
 
-  // Create an updater to keep the store configured appropriately.
-  _updater = new Updater<void, BaseMemcachedStore>(this, std::mem_fun(&BaseMemcachedStore::update_config));
+  if (config_reader != NULL)
+  {
+    // Create an updater to keep the store configured appropriately.
+    _updater = new Updater<void, BaseMemcachedStore>(this, std::mem_fun(&BaseMemcachedStore::update_config));
```
Comment (AlexHockey) [inline src/memcachedstore.cpp]: Should the `_updater` live in the base class. Should it?

**Comment by AlexHockey** [comment]
Looks good, just one comment.

File: src/memcachedstore.cpp (hunk 4 above)
Comment (AlexHockey) [inline src/memcachedstore.cpp]: Should the `_updater` live in `TopologyAwareMemcachedStore`? Does this even work?

File: src/memcachedstore.cpp (hunk 3 above)
Comment (rkd-msw) [inline src/memcachedstore.cpp]: Yep, C++11 brace initialization.

File: src/memcachedstore.cpp (hunk 4 above)
Comment (rkd-msw) [inline src/memcachedstore.cpp]: Yes, it does. See the new commit here - note that I want to talk to you about tombstone_lifetime before merging.

Comment (rkd-msw) [inline src/memcachedstore.cpp]: We agreed that my change to hardcode tombstone_lifetime is fine.

**Pull request merged**